### PR TITLE
 feat: Support gMSAv2 in Windows nodes

### PR DESCRIPTION
**Pull request by AbelHu**
AKS RP will set windowsGmsaPackageUrl if gMSAv2 is enabled. Otherwise it will be set to empty.

### CODE REVIEW

File: parts/windows/windowsconfigfunc.ps1 [hunk 1]
```diff
@@ -196,3 +196,133 @@ function Enable-FIPSMode
...
+
+    # Enable the logging manifest.
+    Write-Log "Importing the CCGEvents manifest file"
+    wevtutil.exe im "$tempInstallPackageFoler\CCGEvents.man"
+    if($LASTEXITCODE) {
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GMSA_IMPORT_CCGEVENTS -ErrorMessage "Failed to import the CCGEvents.man manifest file."
+    }
+
+    # Enable the PowerShell privilege to set the registry permissions.
+    Write-Log "Enabling the PowerShell privilege"
```
Comment (mainred) [inline parts/windows/windowsconfigfunc.ps1:281]: Curious, will setting PowerShell privilege change the behaviors of the following commands outside of `Install-GmsaPlugin`

File: parts/windows/windowsconfigfunc.ps1 [hunk 2]
```diff
@@ -196,3 +196,133 @@ function Enable-FIPSMode
...
+        $acl = $key.GetAccessControl()
+        $acl.SetOwner($owner)
+        $key.SetAccessControl($acl)
+        
+        $key = [Microsoft.Win32.Registry]::LocalMachine.OpenSubKey(
+            $ccgKeyPath,
+            [Microsoft.Win32.RegistryKeyPermissionCheck]::ReadWriteSubTree,
+            [System.Security.AccessControl.RegistryRights]::ChangePermissions)
+        $acl = $key.GetAccessControl()
+        $rule = New-Object System.Security.AccessControl.RegistryAccessRule(
```
Comment (mainred) [inline parts/windows/windowsconfigfunc.ps1:306]: curious, PropagationFlags and AccessControlType are now used?

File: parts/windows/windowsconfigfunc.ps1 [hunk 3]
```diff
@@ -196,3 +196,133 @@ function Enable-FIPSMode
...
+            $owner,
+            [System.Security.AccessControl.RegistryRights]::FullControl,
+            [System.Security.AccessControl.AccessControlType]::Allow)
+        $acl.SetAccessRule($rule)
+        $key.SetAccessControl($acl)
```
Comment (mainred) [inline parts/windows/windowsconfigfunc.ps1:311]: $key.Close()? 
_Closes the key and flushes it to disk if its contents have been modified._

File: parts/windows/windowsconfigfunc.ps1 [hunk 4]
```diff
@@ -196,3 +196,133 @@ function Enable-FIPSMode
...
+        [String] $GmsaPackageUrl
+    )
+
+    if ( $GmsaPackageUrl -eq "" ) {
+        Write-Log "GmsaPackageUrl is not set so skip installing GMSA plugin."
+        return
+    }
+
+    $tempInstallPackageFoler = $env:TEMP
+    $tempPluginZipFile = [Io.path]::Combine($ENV:TEMP, "gmsa.zip")
```
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:257]: This is the name of local temp file. We still use the same naming and format in remote package url

File: parts/windows/windowsconfigfunc.ps1 [hunk 5]
```diff
@@ -196,3 +196,133 @@ function Enable-FIPSMode
...
+
+    Write-Log "Getting the GMSA plugin package"
+    DownloadFileOverHttp -Url $GmsaPackageUrl -DestinationPath $tempPluginZipFile
+    Expand-Archive -Path $tempPluginZipFile -DestinationPath $tempInstallPackageFoler -Force
```
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:261]: We do not see any failure in Expand-Archive. We can do it later if we see such failure.

File: parts/windows/windowsconfigfunc.ps1 (hunk 1 above)
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:281]: It works for the current process to set values under the registry key "System\CurrentControlSet\Control\CCG\COMClasses"

File: parts/windows/windowsconfigfunc.ps1 (hunk 2 above)
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:306]: Cannot get your point. Could you explain your question more?

File: parts/windows/windowsconfigfunc.ps1 (hunk 3 above)
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:311]: It works from my testing. This is only used to run `reg.exe import ` in this process so I think that we do not need to save the data.

**Comment by mainred** [comment]
I compared the powershell script used here with the one provided by Harika Vedati , could you please also invite him to take a look?

File: parts/windows/windowsconfigfunc.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (mainred) [reply parts/windows/windowsconfigfunc.ps1:257]: Sorry you did not answer my question, I mean does the gmsa package has a version? if we have a version we need to differentiate the local cache as well.

File: parts/windows/windowsconfigfunc.ps1 (hunk 2 above)
Comment (mainred) [reply parts/windows/windowsconfigfunc.ps1:314]: my comment is based on the comparision from Harika Vedati's installplugin.ps1.

File: parts/windows/windowsconfigfunc.ps1 (hunk 3 above)
Comment (mainred) [reply parts/windows/windowsconfigfunc.ps1:319]: I guess it's like a file.close(), the decision is given to you.

**Review by mainred — APPROVED** [review]
Not an expert of PowerShell and windows stack. If you are confident with your PR, please go ahead.

File: parts/windows/windowsconfigfunc.ps1 (hunk 3 above)
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:319]: If we need to persistent this change for usage after reboot, we definitely need to call it. But we only use it after enabling privilege and do not want to use it after reboot, I think that it is better not to close it.

File: parts/windows/windowsconfigfunc.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:257]: After offline sync, we can close this.

File: parts/windows/windowsconfigfunc.ps1 (hunk 2 above)
Comment (AbelHu) [reply parts/windows/windowsconfigfunc.ps1:314]: We use different script for the reference.

**Pull request merged**